core/server/common/src/main/java/alluxio/util/ParallelZipUtils.java

```diff
@@ -0,0 +1,181 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Enumeration;
+import java.util.concurrent.CompletionService;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.ExecutorService;
+import java.util.stream.Stream;
+
+/**
+ * Utility methods for working with parallel zip archives.
+ */
+public class ParallelZipUtils {
+  private static final Logger LOG = LoggerFactory.getLogger(ParallelZipUtils.class);
+
+  /**
+   * Creates a zipped archive from the given path in parallel, streaming the data
+   * to the give output path.
+   *
+   * @param dirPath
+   * @param backupPath
+   * @param poolSize
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  public static void compress(Path dirPath, String backupPath, int poolSize)
+      throws IOException, InterruptedException {
+    LOG.info("compress in parallel from path " + dirPath + " to " + backupPath);
+    ExecutorService executor = ExecutorServiceFactories.fixedThreadPool(
+        "parallel-zip-compress-pool", poolSize).create();
+    ParallelScatterZipCreator parallelScatterZipCreator = new ParallelScatterZipCreator(executor);
+
+    try (FileOutputStream fileOutputStream = new FileOutputStream(backupPath);
+         BufferedOutputStream bufferedOutputStream =
+             new BufferedOutputStream(fileOutputStream);
+         ZipArchiveOutputStream zipArchiveOutputStream =
+             new ZipArchiveOutputStream(bufferedOutputStream)) {
+      try (final Stream<Path> stream = Files.walk(dirPath)) {
+        for (Path subPath : stream.collect(toList())) {
+          if (Thread.interrupted()) {
+            throw new InterruptedException();
+          }
+
+          File file = subPath.toFile();
+
+          final InputStreamSupplier inputStreamSupplier = () -> {
+            try {
+              return new FileInputStream(file);
+            } catch (FileNotFoundException e) {
+              LOG.warn("Can't find file when parallel zip, path = {}", subPath);
+              return new NullInputStream(0);
+            }
+          };
+
+          String entryName = dirPath.relativize(subPath).toString();
+          if (file.isDirectory()) {
+            entryName += File.separator;
+          }
+
+          ZipArchiveEntry zipArchiveEntry = new ZipArchiveEntry(entryName);
+          zipArchiveEntry.setMethod(ZipArchiveEntry.DEFLATED);
+
+          parallelScatterZipCreator.addArchiveEntry(zipArchiveEntry, inputStreamSupplier);
+        }
+      }
+
+      parallelScatterZipCreator.writeTo(zipArchiveOutputStream);
+      zipArchiveOutputStream.flush();
+      fileOutputStream.getFD().sync();
```

Comment:
if an IOException is thrown here, will it be wrapped into an `ExecutionException`?